kv/kvserver: TestReplicaClockUpdates failed

[(kv/kvserver).TestReplicaClockUpdates failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2560261&tab=buildLog) on [master@a148c4f0f7ca661d6a44ac54a5f37eb7da037265](https://github.com/cockroachdb/cockroach/commits/a148c4f0f7ca661d6a44ac54a5f37eb7da037265):

Fatal error:
```
panic: Unsafe timestamp cache update! Cannot add timestamp 0.050000001,231? to timestamp cache after evaluating Get ["a",/Min) (resp=(err: <nil>), *roachpb.GetResponse; err=<nil>) with local hlc clock at timestamp 0.000000001,234. The timestamp cache update could be lost on a lease transfer. [recovered]
	panic: Unsafe timestamp cache update! Cannot add timestamp 0.050000001,231? to timestamp cache after evaluating Get ["a",/Min) (resp=(err: <nil>), *roachpb.GetResponse; err=<nil>) with local hlc clock at timestamp 0.000000001,234. The timestamp cache update could be lost on a lease transfer. [recovered]
	panic: Unsafe timestamp cache update! Cannot add timestamp 0.050000001,231? to timestamp cache after evaluating Get ["a",/Min) (resp=(err: <nil>), *roachpb.GetResponse; err=<nil>) with local hlc clock at timestamp 0.000000001,234. The timestamp cache update could be lost on a lease transfer.

```

Stack:
```
goroutine 369791 [running]:
testing.tRunner.func1.1(0x5c2cac0, 0xc03c2e64a0)
	/usr/local/go/src/testing/testing.go:1072 +0x46a
testing.tRunner.func1(0xc01a49ac00)
	/usr/local/go/src/testing/testing.go:1075 +0x636
panic(0x5c2cac0, 0xc03c2e64a0)
	/usr/local/go/src/runtime/panic.go:975 +0x47a
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Send.func1(0xc04b1db028, 0xc04b1db0b8, 0xc026a5b800, 0xc04b1db0b0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_send.go:100 +0x725
panic(0x5c2cac0, 0xc03c2e64a0)
	/usr/local/go/src/runtime/panic.go:975 +0x47a
github.com/cockroachdb/cockroach/pkg/kv/kvserver.checkedTSCacheUpdate.func1(0xc033d91490, 0x1, 0x1, 0x0, 0x0, 0x0, 0x2faf081, 0x1000000e7, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_tscache.go:227 +0x2e5
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).updateTimestampCache(0xc005e68800, 0x7d366e0, 0xc0208d7140, 0xc03541d830, 0xc00aa7e980, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_tscache.go:211 +0x71e
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*endCmds).done(0xc04b1d9910, 0x7d366e0, 0xc0208d7140, 0xc03541d830, 0xc00aa7e980, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica.go:1400 +0x105
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeReadOnlyBatch(0xc005e68800, 0x7d366e0, 0xc0208d7140, 0xc03541d830, 0x0, 0x0, 0xc0276d0ce0, 0x100000001, 0x1, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_read.go:83 +0x645
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeBatchWithConcurrencyRetries(0xc005e68800, 0x7d366e0, 0xc0208d7140, 0xc03541d830, 0x6a2cb30, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:352 +0x690
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendWithRangeID(0xc005e68800, 0x7d366e0, 0xc0468bb7a0, 0x1, 0xc03541d830, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:93 +0x2df
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).Send(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:36
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Send(0xc026a5b800, 0x7d36660, 0xc0001a6010, 0x2faf081, 0x1000000e7, 0x100000001, 0x1, 0x0, 0x1, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_send.go:195 +0x533
github.com/cockroachdb/cockroach/pkg/kv.Wrap.func1(0x7d36660, 0xc0001a6010, 0x2faf081, 0x1000000e7, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:426 +0x155
github.com/cockroachdb/cockroach/pkg/kv.SenderFunc.Send(0xc03a4278a0, 0x7d36660, 0xc0001a6010, 0x2faf081, 0x1000000e7, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:362 +0xcb
github.com/cockroachdb/cockroach/pkg/kv.SendWrappedWith(0x7d36660, 0xc0001a6010, 0x7caaea0, 0xc03a4278a0, 0x2faf081, 0x1000000e7, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:402 +0x195
github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestReplicaClockUpdates.func1(0xc01a49ac00, 0x4e0100)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_replica_test.go:110 +0xa45
github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestReplicaClockUpdates.func2.1(0xc01a49ac00, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_replica_test.go:153 +0x58
github.com/cockroachdb/cockroach/pkg/testutils.RunTrueAndFalse.func1(0xc01a49ac00)
	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/subtest.go:23 +0x6f
testing.tRunner(0xc01a49ac00, 0xc0455ce6a0)
	/usr/local/go/src/testing/testing.go:1123 +0x203
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1168 +0x5bc

```

<details><summary>Log preceding fatal error</summary><p>

```
=== RUN   TestReplicaClockUpdates
    test_log_scope.go:72: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicaClockUpdates411220935
    test_log_scope.go:73: use -show-logs to present logs inline
--- FAIL: TestReplicaClockUpdates (1.42s)
=== RUN   TestReplicaClockUpdates/write=false/synthetic=true
        --- FAIL: TestReplicaClockUpdates/write=false/synthetic=true (0.70s)

```

</p></details>

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestReplicaClockUpdates PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestReplicaClockUpdates.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestReplicaClockUpdates failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2560362&tab=buildLog) on [master@ced6751a28a1803783ed30151b39caeefc45147b](https://github.com/cockroachdb/cockroach/commits/ced6751a28a1803783ed30151b39caeefc45147b):

Fatal error:
```
panic: Unsafe timestamp cache update! Cannot add timestamp 0.050000001,238? to timestamp cache after evaluating Get ["a",/Min) (resp=(err: <nil>), *roachpb.GetResponse; err=<nil>) with local hlc clock at timestamp 0.000000001,241. The timestamp cache update could be lost on a lease transfer. [recovered]
	panic: Unsafe timestamp cache update! Cannot add timestamp 0.050000001,238? to timestamp cache after evaluating Get ["a",/Min) (resp=(err: <nil>), *roachpb.GetResponse; err=<nil>) with local hlc clock at timestamp 0.000000001,241. The timestamp cache update could be lost on a lease transfer. [recovered]
	panic: Unsafe timestamp cache update! Cannot add timestamp 0.050000001,238? to timestamp cache after evaluating Get ["a",/Min) (resp=(err: <nil>), *roachpb.GetResponse; err=<nil>) with local hlc clock at timestamp 0.000000001,241. The timestamp cache update could be lost on a lease transfer.

```

Stack:
```
goroutine 371428 [running]:
testing.tRunner.func1.1(0x5c3fa40, 0xc052c92f60)
	/usr/local/go/src/testing/testing.go:1072 +0x46a
testing.tRunner.func1(0xc03447fc80)
	/usr/local/go/src/testing/testing.go:1075 +0x636
panic(0x5c3fa40, 0xc052c92f60)
	/usr/local/go/src/runtime/panic.go:975 +0x47a
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Send.func1(0xc04c29d028, 0xc04c29d0b8, 0xc05760c000, 0xc04c29d0b0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_send.go:100 +0x725
panic(0x5c3fa40, 0xc052c92f60)
	/usr/local/go/src/runtime/panic.go:975 +0x47a
github.com/cockroachdb/cockroach/pkg/kv/kvserver.checkedTSCacheUpdate.func1(0xc0126ddc70, 0x1, 0x1, 0x0, 0x0, 0x0, 0x2faf081, 0x1000000ee, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_tscache.go:227 +0x2e5
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).updateTimestampCache(0xc05760d000, 0x7d49660, 0xc017a25ad0, 0xc0420a3e60, 0xc0109f3a80, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_tscache.go:211 +0x71e
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*endCmds).done(0xc04c29b910, 0x7d49660, 0xc017a25ad0, 0xc0420a3e60, 0xc0109f3a80, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica.go:1400 +0x105
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeReadOnlyBatch(0xc05760d000, 0x7d49660, 0xc017a25ad0, 0xc0420a3e60, 0x0, 0x0, 0xc035e60f90, 0x100000001, 0x1, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_read.go:83 +0x645
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeBatchWithConcurrencyRetries(0xc05760d000, 0x7d49660, 0xc017a25ad0, 0xc0420a3e60, 0x6a3fab0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:352 +0x690
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendWithRangeID(0xc05760d000, 0x7d49660, 0xc0437d1350, 0x1, 0xc0420a3e60, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:93 +0x2df
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).Send(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:36
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Send(0xc05760c000, 0x7d495e0, 0xc000074140, 0x2faf081, 0x1000000ee, 0x100000001, 0x1, 0x0, 0x1, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_send.go:195 +0x533
github.com/cockroachdb/cockroach/pkg/kv.Wrap.func1(0x7d495e0, 0xc000074140, 0x2faf081, 0x1000000ee, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:426 +0x155
github.com/cockroachdb/cockroach/pkg/kv.SenderFunc.Send(0xc00daaddc0, 0x7d495e0, 0xc000074140, 0x2faf081, 0x1000000ee, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:362 +0xcb
github.com/cockroachdb/cockroach/pkg/kv.SendWrappedWith(0x7d495e0, 0xc000074140, 0x7cbde20, 0xc00daaddc0, 0x2faf081, 0x1000000ee, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:402 +0x195
github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestReplicaClockUpdates.func1(0xc03447fc80, 0x4e0100)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_replica_test.go:110 +0xa45
github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestReplicaClockUpdates.func2.1(0xc03447fc80, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_replica_test.go:153 +0x58
github.com/cockroachdb/cockroach/pkg/testutils.RunTrueAndFalse.func1(0xc03447fc80)
	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/subtest.go:23 +0x6f
testing.tRunner(0xc03447fc80, 0xc049b76680)
	/usr/local/go/src/testing/testing.go:1123 +0x203
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1168 +0x5bc

```

<details><summary>Log preceding fatal error</summary><p>

```
=== RUN   TestReplicaClockUpdates
    test_log_scope.go:72: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicaClockUpdates506188776
    test_log_scope.go:73: use -show-logs to present logs inline
--- FAIL: TestReplicaClockUpdates (1.48s)
=== RUN   TestReplicaClockUpdates/write=false/synthetic=true
        --- FAIL: TestReplicaClockUpdates/write=false/synthetic=true (0.75s)

```

</p></details>

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestReplicaClockUpdates PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestReplicaClockUpdates.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestReplicaClockUpdates failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2560442&tab=buildLog) on [master@b2ac447d23c97332b705e288c055d63119fb8d28](https://github.com/cockroachdb/cockroach/commits/b2ac447d23c97332b705e288c055d63119fb8d28):

Fatal error:
```
panic: Unsafe timestamp cache update! Cannot add timestamp 0.050000001,237? to timestamp cache after evaluating Get ["a",/Min) (resp=(err: <nil>), *roachpb.GetResponse; err=<nil>) with local hlc clock at timestamp 0.000000001,240. The timestamp cache update could be lost on a lease transfer. [recovered]
	panic: Unsafe timestamp cache update! Cannot add timestamp 0.050000001,237? to timestamp cache after evaluating Get ["a",/Min) (resp=(err: <nil>), *roachpb.GetResponse; err=<nil>) with local hlc clock at timestamp 0.000000001,240. The timestamp cache update could be lost on a lease transfer. [recovered]
	panic: Unsafe timestamp cache update! Cannot add timestamp 0.050000001,237? to timestamp cache after evaluating Get ["a",/Min) (resp=(err: <nil>), *roachpb.GetResponse; err=<nil>) with local hlc clock at timestamp 0.000000001,240. The timestamp cache update could be lost on a lease transfer.

```

Stack:
```
goroutine 373152 [running]:
testing.tRunner.func1.1(0x5c3fa00, 0xc04dec4180)
	/usr/local/go/src/testing/testing.go:1072 +0x46a
testing.tRunner.func1(0xc0687d8000)
	/usr/local/go/src/testing/testing.go:1075 +0x636
panic(0x5c3fa00, 0xc04dec4180)
	/usr/local/go/src/runtime/panic.go:975 +0x47a
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Send.func1(0xc0313cb028, 0xc0313cb0b8, 0xc05da11800, 0xc0313cb0b0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_send.go:100 +0x725
panic(0x5c3fa00, 0xc04dec4180)
	/usr/local/go/src/runtime/panic.go:975 +0x47a
github.com/cockroachdb/cockroach/pkg/kv/kvserver.checkedTSCacheUpdate.func1(0xc0665c3db8, 0x1, 0x1, 0x0, 0x0, 0x0, 0x2faf081, 0x1000000ed, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_tscache.go:227 +0x2e5
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).updateTimestampCache(0xc066614800, 0x7d49660, 0xc04670df20, 0xc06e98b8c0, 0xc062366980, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_tscache.go:211 +0x71e
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*endCmds).done(0xc0313c9910, 0x7d49660, 0xc04670df20, 0xc06e98b8c0, 0xc062366980, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica.go:1400 +0x105
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeReadOnlyBatch(0xc066614800, 0x7d49660, 0xc04670df20, 0xc06e98b8c0, 0x0, 0x0, 0xc069616660, 0x100000001, 0x1, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_read.go:83 +0x645
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeBatchWithConcurrencyRetries(0xc066614800, 0x7d49660, 0xc04670df20, 0xc06e98b8c0, 0x6a3fa70, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:352 +0x690
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendWithRangeID(0xc066614800, 0x7d49660, 0xc06a3a84e0, 0x1, 0xc06e98b8c0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:93 +0x2df
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).Send(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:36
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Send(0xc05da11800, 0x7d495e0, 0xc00020c010, 0x2faf081, 0x1000000ed, 0x100000001, 0x1, 0x0, 0x1, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_send.go:195 +0x533
github.com/cockroachdb/cockroach/pkg/kv.Wrap.func1(0x7d495e0, 0xc00020c010, 0x2faf081, 0x1000000ed, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:426 +0x155
github.com/cockroachdb/cockroach/pkg/kv.SenderFunc.Send(0xc04230bca0, 0x7d495e0, 0xc00020c010, 0x2faf081, 0x1000000ed, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:362 +0xcb
github.com/cockroachdb/cockroach/pkg/kv.SendWrappedWith(0x7d495e0, 0xc00020c010, 0x7cbde20, 0xc04230bca0, 0x2faf081, 0x1000000ed, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:402 +0x195
github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestReplicaClockUpdates.func1(0xc0687d8000, 0x4e0100)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_replica_test.go:110 +0xa45
github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestReplicaClockUpdates.func2.1(0xc0687d8000, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_replica_test.go:153 +0x58
github.com/cockroachdb/cockroach/pkg/testutils.RunTrueAndFalse.func1(0xc0687d8000)
	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/subtest.go:23 +0x6f
testing.tRunner(0xc0687d8000, 0xc04f60ed20)
	/usr/local/go/src/testing/testing.go:1123 +0x203
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1168 +0x5bc

```

<details><summary>Log preceding fatal error</summary><p>

```
=== RUN   TestReplicaClockUpdates
    test_log_scope.go:72: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicaClockUpdates715642231
    test_log_scope.go:73: use -show-logs to present logs inline
--- FAIL: TestReplicaClockUpdates (1.40s)
=== RUN   TestReplicaClockUpdates/write=false/synthetic=true
        --- FAIL: TestReplicaClockUpdates/write=false/synthetic=true (0.72s)

```

</p></details>

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestReplicaClockUpdates PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestReplicaClockUpdates.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestReplicaClockUpdates failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2560509&tab=buildLog) on [master@04063a9f699d37dd95696210a52e5c6d8f536b86](https://github.com/cockroachdb/cockroach/commits/04063a9f699d37dd95696210a52e5c6d8f536b86):

Fatal error:
```
panic: Unsafe timestamp cache update! Cannot add timestamp 0.050000001,231? to timestamp cache after evaluating Get ["a",/Min) (resp=(err: <nil>), *roachpb.GetResponse; err=<nil>) with local hlc clock at timestamp 0.000000001,234. The timestamp cache update could be lost on a lease transfer. [recovered]
	panic: Unsafe timestamp cache update! Cannot add timestamp 0.050000001,231? to timestamp cache after evaluating Get ["a",/Min) (resp=(err: <nil>), *roachpb.GetResponse; err=<nil>) with local hlc clock at timestamp 0.000000001,234. The timestamp cache update could be lost on a lease transfer. [recovered]
	panic: Unsafe timestamp cache update! Cannot add timestamp 0.050000001,231? to timestamp cache after evaluating Get ["a",/Min) (resp=(err: <nil>), *roachpb.GetResponse; err=<nil>) with local hlc clock at timestamp 0.000000001,234. The timestamp cache update could be lost on a lease transfer.

```

Stack:
```
goroutine 368619 [running]:
testing.tRunner.func1.1(0x5c2ca80, 0xc065f4fd90)
	/usr/local/go/src/testing/testing.go:1072 +0x46a
testing.tRunner.func1(0xc05b44e000)
	/usr/local/go/src/testing/testing.go:1075 +0x636
panic(0x5c2ca80, 0xc065f4fd90)
	/usr/local/go/src/runtime/panic.go:975 +0x47a
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Send.func1(0xc06859b028, 0xc06859b0b8, 0xc05a89d800, 0xc06859b0b0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_send.go:100 +0x725
panic(0x5c2ca80, 0xc065f4fd90)
	/usr/local/go/src/runtime/panic.go:975 +0x47a
github.com/cockroachdb/cockroach/pkg/kv/kvserver.checkedTSCacheUpdate.func1(0xc0657a73c0, 0x1, 0x1, 0x0, 0x0, 0x0, 0x2faf081, 0x1000000e7, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_tscache.go:227 +0x2e5
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).updateTimestampCache(0xc05ed06000, 0x7d366e0, 0xc017a297d0, 0xc034a7a6c0, 0xc0676e4000, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_tscache.go:211 +0x71e
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*endCmds).done(0xc06c059910, 0x7d366e0, 0xc017a297d0, 0xc034a7a6c0, 0xc0676e4000, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica.go:1400 +0x105
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeReadOnlyBatch(0xc05ed06000, 0x7d366e0, 0xc017a297d0, 0xc034a7a6c0, 0x0, 0x0, 0xc0606bb580, 0x100000001, 0x1, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_read.go:83 +0x645
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeBatchWithConcurrencyRetries(0xc05ed06000, 0x7d366e0, 0xc017a297d0, 0xc034a7a6c0, 0x6a2caf0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:352 +0x690
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendWithRangeID(0xc05ed06000, 0x7d366e0, 0xc05e68cc90, 0x1, 0xc034a7a6c0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:93 +0x2df
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).Send(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:36
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Send(0xc05a89d800, 0x7d36660, 0xc000074140, 0x2faf081, 0x1000000e7, 0x100000001, 0x1, 0x0, 0x1, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_send.go:195 +0x533
github.com/cockroachdb/cockroach/pkg/kv.Wrap.func1(0x7d36660, 0xc000074140, 0x2faf081, 0x1000000e7, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:426 +0x155
github.com/cockroachdb/cockroach/pkg/kv.SenderFunc.Send(0xc067011ba0, 0x7d36660, 0xc000074140, 0x2faf081, 0x1000000e7, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:362 +0xcb
github.com/cockroachdb/cockroach/pkg/kv.SendWrappedWith(0x7d36660, 0xc000074140, 0x7caaea0, 0xc067011ba0, 0x2faf081, 0x1000000e7, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:402 +0x195
github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestReplicaClockUpdates.func1(0xc05b44e000, 0x4e0100)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_replica_test.go:110 +0xa45
github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestReplicaClockUpdates.func2.1(0xc05b44e000, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_replica_test.go:153 +0x58
github.com/cockroachdb/cockroach/pkg/testutils.RunTrueAndFalse.func1(0xc05b44e000)
	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/subtest.go:23 +0x6f
testing.tRunner(0xc05b44e000, 0xc06917cd60)
	/usr/local/go/src/testing/testing.go:1123 +0x203
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1168 +0x5bc

```

<details><summary>Log preceding fatal error</summary><p>

```
=== RUN   TestReplicaClockUpdates
    test_log_scope.go:72: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicaClockUpdates069795429
    test_log_scope.go:73: use -show-logs to present logs inline
--- FAIL: TestReplicaClockUpdates (1.50s)
=== RUN   TestReplicaClockUpdates/write=false/synthetic=true
        --- FAIL: TestReplicaClockUpdates/write=false/synthetic=true (0.71s)

```

</p></details>

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestReplicaClockUpdates PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestReplicaClockUpdates.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/kv/kvserver/replica_tscache.go
+++ b/pkg/kv/kvserver/replica_tscache.go
@@ -223,7 +223,7 @@ func checkedTSCacheUpdate(
 	pErr *roachpb.Error,
 ) func(roachpb.Key, roachpb.Key, hlc.Timestamp, uuid.UUID) {
 	return func(start, end roachpb.Key, ts hlc.Timestamp, txnID uuid.UUID) {
-		if now.Less(ts) {
+		if now.Less(ts) && !ts.Synthetic {
 			panic(fmt.Sprintf("Unsafe timestamp cache update! Cannot add timestamp %s to timestamp "+
 				"cache after evaluating %v (resp=%v; err=%v) with local hlc clock at timestamp %s. "+
 				"The timestamp cache update could be lost on a lease transfer.", ts, ba, br, pErr, now))

